### PR TITLE
Pin to taskbar from settings on Windows

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/24509
fix https://github.com/brave/brave-browser/issues/24580

Pin current profile's shortcut to taskbar when user set Brave as
a default via settings.
If taskbar has any brave shortcut, we allow to pin only once.

To pin to taskbar, shortcut file is created in temp folder instead
of in the desktop folder

Resolves 

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

**_NOTE: This only works for Win10 or later version._**

### Case 1 for verifying https://github.com/brave/brave-browser/issues/24580
1. Clean existing profile and download installer
2. Run installer(System level or user-level install)
3. Allow set as a default browser from first run dialog
4. Check additional shortcut is not created in desktop folder. (Maybe one shortcut is there. it's created by installer)

### Case 2
1. Unpin from taskbar if pinned before and set non-brave as a default browser
2. Launch browser with clean profile
3. Load brave://settings/ and set as default browser
4. Check brave is pinned 
5. Shutdown Brave and launch from taskbar shortcut
6. Check brave is launched properly

### Case 3
1. Unpin from tasbar if pinned before and set non-brave as a default browser
2. Launch browser and create another profile
3. Load brave://settings from newly created profile and set as a default browser
4. Check new profile's shortcut is pinned
5. Shutdown Brave and launch from taskbar shortcut
6. Check second profile's window is launched properly

### Case 4
1. Repeat above case 2's all steps
2. Set non-brave as a default browser
3. Create another profile
4. Load brave://settings from newly created profile and set as a default browser
5. Check new profile's shortcut is not pinned because we have already another one

### CODE REVIEW

**Comment by simonhong** [comment]
After pinned, another icon is created during the run time instead of using pinned icon in the taskbar. Investigating...
image

**Comment by simonhong** [comment]
When pin to taskbar, an existing shortcut should be passed to pin method.
If a shortcut is not existed for current profile(or default one), it should be created.
Current PR, creates it in Desktop folder and deletes it after pinned. 
Doing in temp dir instead of Desktop folder would be better to hide intermediate file creation from users.

File: browser/brave_shell_integration_win.cc [hunk 1]
```diff
@@ -123,51 +191,192 @@ absl::optional<bool> IsShortcutPinnedWin10(const base::FilePath& shortcut) {
...
+
+  // TODO(simonhong): Support win7/8
+  if (base::win::GetVersion() < base::win::Version::WIN10_RS5)
+    return false;
+
+  return true;
+}
+
+bool ForceToPinToTaskbar() {
+  // We pin current profile's shortcut only once even if taskbar has other
```
Comment (goodov) [inline browser/brave_shell_integration_win.cc:330]: `DCHECK_CURRENTLY_ON(content::BrowserThread::UI);`?

File: browser/brave_shell_integration_win.cc [hunk 2]
```diff
@@ -123,51 +191,192 @@ absl::optional<bool> IsShortcutPinnedWin10(const base::FilePath& shortcut) {
...
+    profile_attrs.push_back(
+        std::make_tuple(entry->GetPath(), entry->GetName(),
+                        win::GetAppUserModelIdForBrowser(entry->GetPath())));
+  }
+
   base::ThreadPool::CreateCOMSTATaskRunner({base::MayBlock()})
-      ->PostTask(FROM_HERE, base::BindOnce(&PinToTaskbarImpl));
+      ->PostTaskAndReplyWithResult(
+          FROM_HERE,
+          base::BindOnce(&HasTaskbarAnyPinnedBraveShortcuts, profile_attrs),
```
Comment (goodov) [inline browser/brave_shell_integration_win.cc:369]: `std::move(profile_attrs)`

File: browser/brave_shell_integration_win.h [hunk 3]
```diff
@@ -6,9 +6,15 @@
...
 
+class Profile;
+
 namespace shell_integration::win {
 
-void PinToTaskbar();
+// Pin profile-specific shortcut when |profile| is non-null.
+// Set true to |check_existing_pinned_shortcut| if want to pin
+// only when taskbar doesn't have any pinned brave shortcuts.
+void PinToTaskbar(Profile* profile = nullptr,
```
Comment (goodov) [inline browser/brave_shell_integration_win.h:16]: why ` = nullptr`? is this a valid case?

Comment (simonhong) [reply browser/brave_shell_integration_win.h:16]: It's valid case.
`PinToTaskbar()` could be called w/o any args when we want to pin default shortcut to taskbar
or active profile could not be available at the early stage like at FirstRun dialog.

**Review by simonhong — COMMENTED** [review]
@goodov Addressed. Thanks! PTAL again.

**Review by goodov — APPROVED** [review]

File: browser/brave_shell_integration_win.cc [hunk 4]
```diff
@@ -10,21 +10,49 @@
...
 
+class ScopedShortcutFile {
+ public:
+  explicit ScopedShortcutFile(const base::FilePath& file_path)
+      : file_path_(file_path) {
+    DCHECK(!file_path_.empty());
+  }
+  ScopedShortcutFile(const ScopedShortcutFile&) = delete;
+  ScopedShortcutFile& operator=(const ScopedShortcutFile&) = delete;
+  virtual ~ScopedShortcutFile() { base::DeleteFile(file_path_); }
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/brave_shell_integration_win.cc]: Done. thanks for review! 👍🏼

**Pull request merged**